src/main/java/com/aws/greengrass/deployment/DeploymentDocumentDownloader.java

```diff
@@ -0,0 +1,218 @@
+                                        HttpClientProvider httpClientProvider) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+        this.thingNameTopic = deviceConfiguration.getThingName();
+        this.httpClientProvider = httpClientProvider;
+    }
+
+    /**
+     * Download the full deployment document stored from greengrass cloud.
+     *
+     * @param deploymentId deployment id
+     * @return DeploymentDocument
+     * @throws DeploymentDocumentDownloadException if failed to download the full deployment document.
+     */
+    public DeploymentDocument download(String deploymentId) throws DeploymentDocumentDownloadException {
+        // TODO [Next Revision] Support downloading retrials for intermittent network and temporary GG data plane
+        //  service error with RetryUtils
+
+        // 1. Get url, digest, and algorithm by calling gg data plane
+        GetDeploymentConfigurationResponse response = getDeploymentConfiguration(deploymentId);
+
+        String preSignedUrl = response.preSignedUrl();
+        String algorithm = response.integrityCheck().algorithm();
+        String digest = response.integrityCheck().digest();
+
+        validate(preSignedUrl, algorithm, digest);
+
+        // 2. Download configuration and check integrity
+        String configurationInString = downloadFromUrl(deploymentId, preSignedUrl);
+        checkIntegrity(digest, configurationInString);
+
+        // 3. deserialize and convert to device model
+        Configuration configuration = deserializeDeploymentDoc(configurationInString);
+        return DeploymentDocumentConverter.convertFromDeploymentConfiguration(configuration);
+
+    }
+
+    private String downloadFromUrl(String deploymentId, String preSignedUrl)
+            throws DeploymentDocumentDownloadException {
+        HttpExecuteRequest executeRequest = HttpExecuteRequest.builder()
+                .request(SdkHttpFullRequest.builder().uri(URI.create(preSignedUrl)).method(SdkHttpMethod.GET).build())
+                .build();
+
+        logger.atInfo().kv("DeploymentId", deploymentId)
+                .log("Making HTTP request to the presigned url");  // url is not logged for security concerns
+
+        try (SdkHttpClient client = httpClientProvider.getSdkHttpClient()) {
+
+            HttpExecuteResponse executeResponse;
+            try {
+                executeResponse = client.prepareRequest(executeRequest).call();
+            } catch (IOException e) {
+                throw new DeploymentDocumentDownloadException("I/O error when making HTTP request with presigned url.",
+                        e);
+            }
+
+            validateHttpExecuteResponse(executeResponse);
+
+            try (InputStream in = executeResponse.responseBody().get()) { // checked in validateResponse
+                // load directly into memory because it will be used for resolving configuration
```

Comment:
The limit is 10MB, which is from `CreateDeployment` API, where customer's entire payload size can't exceed 10MB. I put in the comment before but might have deleted it in the new revision.

It's a great suggestion and plus one on we should check the size before doing a get. Especially because we also want to be more protective and not purely rely on the 10MB contract.

Reply:
added size limit check before downloading the file.